providers/siw/siw.c

```diff
@@ -0,0 +1,941 @@
+	atomic_store_explicit((a), b, memory_order_relaxed)
+
+static int siw_notify_cq(struct ibv_cq *ibcq, int solicited)
+{
+	struct siw_cq *cq = cq_base2siw(ibcq);
+	atomic_ulong *notifyp = (atomic_ulong *)&cq->ctrl->notify;
+	int rv = 0;
+
+	if (solicited)
+		_store_mmapped(notifyp, SIW_NOTIFY_SOLICITED);
+	else
+		_store_mmapped(notifyp, SIW_NOTIFY_SOLICITED |
+			       SIW_NOTIFY_NEXT_COMPLETION);
+	return rv;
+}
+
+static enum siw_opcode map_send_opcode(enum ibv_wr_opcode ibv_op)
+{
+	switch (ibv_op) {
+	case IBV_WR_SEND:
+		return SIW_OP_SEND;
+	case IBV_WR_RDMA_WRITE:
+		return SIW_OP_WRITE;
+	case IBV_WR_RDMA_READ:
+		return SIW_OP_READ;
+	default:
+		printf("libsiw: op %d not supported\n", ibv_op);
+	}
+	return SIW_NUM_OPCODES + 1;
+}
+
+static inline uint16_t map_send_flags(int ibv_flags)
+{
+	uint16_t flags = SIW_WQE_VALID;
+
+	if (ibv_flags & IBV_SEND_SIGNALED)
+		flags |= SIW_WQE_SIGNALLED;
+	if (ibv_flags & IBV_SEND_SOLICITED)
+		flags |= SIW_WQE_SOLICITED;
+	if (ibv_flags & IBV_SEND_INLINE)
+		flags |= SIW_WQE_INLINE;
+	if (ibv_flags & IBV_SEND_FENCE)
+		flags |= SIW_WQE_READ_FENCE;
+
+	return flags;
+}
+
+static inline int push_send_wqe(struct ibv_send_wr *base_wr,
+				struct siw_sqe *siw_sqe, int sig_all)
+{
+	uint32_t flags = map_send_flags(base_wr->send_flags);
+	atomic_ushort *fp = (atomic_ushort *)&siw_sqe->flags;
+
+	siw_sqe->id = base_wr->wr_id;
+	siw_sqe->num_sge = base_wr->num_sge;
+	siw_sqe->raddr = base_wr->wr.rdma.remote_addr;
+	siw_sqe->rkey = base_wr->wr.rdma.rkey;
+
+	siw_sqe->opcode = map_send_opcode(base_wr->opcode);
+	if (siw_sqe->opcode > SIW_NUM_OPCODES) {
+		if (siw_debug)
+			printf("libsiw: opcode %d unsupported\n",
+			       base_wr->opcode);
+		return -EINVAL;
+	}
+	if (sig_all)
+		flags |= SIW_WQE_SIGNALLED;
+
+	if (flags & SIW_WQE_INLINE) {
+		char *data = (char *)&siw_sqe->sge[1];
+		int bytes = 0, i = 0;
+
+		/* Allow more than SIW_MAX_SGE, since content copied here */
+		while (i < base_wr->num_sge) {
+			bytes += base_wr->sg_list[i].length;
+			if (bytes > (int)SIW_MAX_INLINE) {
+				if (siw_debug)
+					printf("libsiw: inline data: %d:%d\n",
+					       bytes, (int)SIW_MAX_INLINE);
+				return -EINVAL;
+			}
+			memcpy(data,
+			       (void *)(unsigned long)base_wr->sg_list[i].addr,
+			       base_wr->sg_list[i].length);
+			data += base_wr->sg_list[i++].length;
+		}
+		siw_sqe->sge[0].length = bytes;
+
+	} else if (base_wr->num_sge == 1) {
+		siw_sqe->sge[0].laddr = base_wr->sg_list[0].addr;
+		siw_sqe->sge[0].length = base_wr->sg_list[0].length;
+		siw_sqe->sge[0].lkey = base_wr->sg_list[0].lkey;
+	} else if (base_wr->num_sge && base_wr->num_sge <= SIW_MAX_SGE)
+		/* this assumes same layout of siw and base SGE */
+		memcpy(siw_sqe->sge, base_wr->sg_list,
+		       siw_sqe->num_sge * sizeof(struct ibv_sge));
+	else
+		return -EINVAL;
+
+	_store_mmapped(fp, flags);
+
+	return 0;
+}
+
+static int siw_db(struct siw_qp *qp)
+{
+	int rv;
+
+	rv = write(qp->base_qp.context->cmd_fd, &qp->db_req,
```

Comment:
Huh? Drivers should not be calling write on the cmd_fd - what is this trying to do?

Reply:
It rings the SQ DB in case a new SQE got written to the mmapped queue
and the siw kernel thread is likely idling. Basically, it pushes an empty
SQE as efficient as possible. It is what also librxe does as a HW DB surrogate.

Reply:
Oh that is gross.

If you want to signal the kernel cheaply (or vice versa) then please use an eventfd. Have userspace create one and then pass the FD # into the kernel during get_context. write() from userspace on the event fd can wake up the kernel.

See vfio_virqfd_enable for how this works.

Reply:
I think the sole reason to allow for IB_USER_VERBS_CMD_POST_SEND
is to support sw rdma userlibs like rxe and siw. Since both have
mmapped SQ's, nothing is to be passed down during post_send,
except the fact the SQ got updated. It would be probably better
to drop that IB_USER_VERBS_CMD_POST_SEND support, and replace it
with a IB_USER_VERBS_CMD_DB_SEND, which just transfers the QP
handle, and is forwarded by ib_core largely uninterpreted.

Besides the effort of wrapping my brain around the whole
vfio_virqfd thing - do we really want to waste another fd just
for passing down that one piece of information? 

So I would be very happy if we could leave it as is for
now (same for rxe), and I'd volunteer to introduce that
IB_USER_VERBS_CMD_DB_SEND for a next release.

Reply:
post_send was not intended to be used like this, it is some rxe abuse to do so. hfi/qib use it more sensibly.

eventfd will be much faster than a IB_USER_VERBS_CMD_DB_SEND - there is lots of overheard on the path. The design of the new api was always to have companion fds for the high speed path.

The whole vfio_virqqfd thing is pretty simple and gives you exactly what siw needs..

We want to look at the rdma-core side before merging kernel patches specifically because it is easier to see ABI abuses  like this from userspace :)

Reply:
So I tried a little this stuff. To my understanding, it would have to be per QP?? What I get in the kernel is just being called on a re-registered handler with an opaque argument, which I can define during vfio_virqfd_enable().
That single argument doesnt help me much if it is not the QP itself. Writing to the eventfd is limited to an 8 bytes buffer which adds the contained value to some bogus kernel counter I can read afterwards from userland. I do not get that number (which would be the QP number in my case) down there in the driver.

If I do not did  it wrong -- I think one fd per QP is rather heavy stuff. one fd per context would be ok.

Reply:
Yeah, you have that right. Sigh. OK let us leave it be then.

Reply:
that's unfortunate. why do these guys do not pass the written value along? I'd be down there at my context in a little more than 200ns (measured on a simple i5 box). that would be pretty fast....the post_send command seem to take ages compared against that ;)